feat(pairing): pairing manager API integration

flotter opened the pull request:
PR will be migrated to Pebble and properly commented once pairing manager is merged. This provides a delta view as the PR sits on top.

flotter committed:
feat(pairing): add pairing api support

anpep commented on internals/daemon/api_pairing.go:
```diff
@@ -0,0 +1,58 @@
+				return BadRequest("cannot pair client: %v", err)
+			}
+		default:
+			return BadRequest("cannot pair using %v transport type: not supported", transport.String())
```
nit:
```suggestion
			return BadRequest("cannot pair using %q transport type: not supported", transport)
```

flotter replied:
Fixed. Resolving.

flotter commented on internals/daemon/api_pairing.go:
```diff
@@ -0,0 +1,58 @@
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package daemon
```
Add API docs (JSON file needs updating)

flotter commented on client/client.go:
```diff
@@ -134,9 +135,16 @@ type Config struct {
+	// certificate as needed for mutual TLS to authenticate the
+	// client on the server side. If not supplied, the default
+	// handler will return an error.
+	TLSClientCertificate func() (*tls.Certificate, error)
```
TLSServerVerify *x509.Certificate  ->

client.TLSServerCertificate() (X509.Certificate) <-----

TLSClientCertificate *tls.Certificate ->

flotter commented on internals/overlord/tlsstate/manager.go:
```diff
@@ -137,13 +137,23 @@ func (m *TLSManager) ListenConfig() *tls.Config {
+		NextProtos:       []string{"h2", "http/1.1"},
+		MinVersion:       tls.VersionTLS13,
+		GetCertificate:   m.GetCertificate,
+		ClientAuth:       tls.RequestClientCert,
```
Make sure client transcript is verified against the client certificate with this option.

flotter replied:
This is true. Verified in source code.

paul-rodriguez commented on client/client.go:
```diff
@@ -134,9 +150,44 @@ type Config struct {
+	// is trusted and pinned, future TLS connections must supply the
+	// TLSServerIDCert field, and leave TLSServerFingerprint empty.
+	TLSServerInsecure bool
+	// TLSClientIDCert must hold the client identity certificate when
+	// using a TLS connection to the server. This field must be nil if
+	// a non-TLS based transport is used.
+	TLSClientIDCert *tls.Certificate
```
Is this a requirement even without mTLS? Or are we already enforcing mTLS?

flotter replied:
Discussed with Paul.

paul-rodriguez commented on client/client.go:
```diff
@@ -339,12 +379,20 @@ func (rq *defaultRequester) Do(ctx context.Context, opts *RequestOptions) (*Requ
+	// If this is a TLS connection, extract the server identity certificate
+	// which is placed after the TLS certificate.
+	if httpResp.TLS != nil && len(httpResp.TLS.PeerCertificates) == expectedServerCertCount {
+		idCert = httpResp.TLS.PeerCertificates[1]
```
It seems a bit brittle that we rely on the order of certificates here. Would it be a lot of extra effort to read the name or some other part of the certificate to instead scan for the correct one?

flotter replied:
As discussed, will add some documentation /  commentary to make this clearer.

paul-rodriguez commented on client/client.go:
```diff
@@ -630,6 +685,66 @@ func newDefaultRequester(client *Client, opts *Config) (*defaultRequester, error
+		return nil
+	}
+
+	return errors.New("cannot verify server: see TLS config options")
```
```suggestion
	return errors.New("cannot verify server: missing TLS config option")
```

flotter replied:
Fixed.

paul-rodriguez commented on internals/overlord/state/identities.go:
```diff
@@ -353,26 +355,40 @@ func (s *State) Identities() map[string]*Identity {
+			}
+		}
+		// If a client certificate is provided, but did not match, we bail.
+		return nil
```
I think it would be worth also returning an error when the returned Identity is nil.

anpep replied:
I think we should distinguish between errors and no identity found (`nil`), so I would make no changes here

paul-rodriguez commented on internals/overlord/state/identities_test.go:
```diff
@@ -41,6 +41,16 @@ jwXVTUH4HLpbhK0RAaEPOL4h5jm36CrWTkxzpbdCrIu4NgPLQKJ6Cw==
+VQQKDAtFeGFtcGxlIE9yZzAqMAUGAytlcAMhAIlut+P3huKtFK439Ap+7U4Bv4r2
+DY3fLYnfNEcrXTdLMAUGAytlcANBAEhUiFSTNuCuu2rc4pqGwXYGEtEFqRZDZwYe
+mHLySscsVEgGwncFhL/9UW5iZl/tO/o+WiyVd/K4Vk0Yrp6uggA=
+-----END CERTIFICATE-----
```
These two certificates appear to be used as "valid" and "invalid" respectively, so I suggest to name them that way too.

flotter replied:
Fixed. Resolving.

paul-rodriguez reviewed (commented):
Ended 1st pass, left a couple of minor comments

benhoyt committed:
fix: don't run "start" tasks again, to avoid spurious restarts (#709)

If a service is configured with "on-failure: shutdown" and it fails,
Pebble will shut down, but then when Pebble runs again it'll try to redo
that "start" task and immediately shut down again. The task is in status
3 (DoingStatus).

This change fixes that bug: it'll skip running "start" tasks again, if
they've been started once before (from a previous Pebble run, for
example).

This PR also adds an integration test (regression test) for that.

Fixes #527.

anpep commented on client/client.go:
```diff
@@ -630,6 +685,66 @@ func newDefaultRequester(client *Client, opts *Config) (*defaultRequester, error
+	//
+	// state.PeerCertificates[0] - server TLS certificate
+	// state.PeerCertificates[1] - server Identity certificate (root CA)
+	certCount := len(state.PeerCertificates)
+	if certCount != expectedServerCertCount {
+		return fmt.Errorf("cannot find identity certificate: expected %d certificates, got %d", expectedServerCertCount, certCount)
+	}
```
```suggestion
	if len(state.PeerCertificates) != expectedServerCertCount {
		return fmt.Errorf("cannot find identity certificate: expected %d certificates, got %d", expectedServerCertCount, certCount)
	}
```

flotter replied:
certCount is used twice. Will stick with the current approach.

flotter committed:
feat(pairingstate): add client pairing manager [1/2] (#702)

The pairing manager enables and processes incoming mTLS client pairing
requests on the ```/v1/pairing``` HTTP end-point.

The pairing manager does not by itself decide when the pairing window
should be enabled. The ```EnablePairing``` call will be made by another
manager (called the trigger manager) that deals with event bindings,
such as mapping a button press to an enable pairing request. The trigger
manager will be added at a later stage.

This pull-request focuses only on the Pairing Manager itself,
implementing the following functionality (KO093, KO088):

```go
// PairingEnabled returns whether pairing is currently enabled.
func (m *PairingManager) PairingEnabled() bool {...}
```

```go
// EnablePairing requests the pairing manager to enable the pairing window.
func (m *PairingManager) EnablePairing(timeout time.Duration) error {...}
```

```go
// PairMTLS adds a certificate based client identity with admin permissions to
// the identity subsystem. A pairing request always closes the pairing window.
func (m *PairingManager) PairMTLS(clientCert *x509.Certificate) error {...}
```

Integration of the pairing manager into the HTTPS server and access
checks will follow in #708

flotter committed:
Code Review group 1

flotter commented:
PR migrated to Pebble.
